inbox contract v2 (#3)
ethstorage/specs: specs/experimental/inbox-contract.md

    - [How `op-node` knows the canonical batch inbox](#how-op-node-knows-the-canonical-batch-inbox)
    - [How `op-batcher` knows canonical batch inbox](#how-op-batcher-knows-canonical-batch-inbox)
  - [Upgrade](#upgrade)
  - [Reference Implementation](#reference-implementation)

## Motivation

The batch inbox is currently an Externally Owned Account (EOA), which has both advantages and disadvantages:

Advantages:
- Low submission gas cost due to the absence of onchain execution.
- Verification logic is moved offchain to the derivation part, protected by a fault dispute game with a correct absolute prestate.

Disadvantage:
- Onchain verification is not possible.

This specification aims to allow the batch inbox to be a contract, enabling customized batch submission conditions such as:
- Requiring the batch transaction to be signed by a quorum of sequencers in a decentralized sequencing network; or
- Mandating that the batch transaction call a BLOB storage contract (e.g., EthStorage) with a long-term storage fee, which is then distributed to data nodes that prove BLOB storage over time.

## How It Works

The integration process consists of three primary components:

> 1. Replacement of the [`BatchInboxAddress`](https://github.com/ethereum-optimism/optimism/blob/db107794c0b755bc38a8c62f11c49320c95c73db/op-chain-ops/genesis/config.go#L77) with an inbox contract: The existing `BatchInboxAddress`, which currently points to an Externally Owned Account (EOA), will be replaced by a smart contract. This new inbox contract will be responsible for verifying and enforcing batch submission conditions.

Reviewer: The term `replacement` may be misleading as the inbox address can be initialized in the Rollup config (so no replacement is needed).

Author: Good point.

Author: The term `replacement` is now gone: https://github.com/ethstorage/specs/pull/3/commits/2f22e44289838c26da04bdfb16fc419e8a84b459

2. Modification of the `op-node` derivation process: The `op-node` will be updated to exclude failed batch transactions during the derivation process. This change ensures that only successfully executed batch transactions are processed and included in the derived state. 
3. Modification of the op-batcher submission process: The op-batcher will be updated to [call `recordFailedTx`](https://github.com/blockchaindevsh/optimism/blob/02e3b7248f1b590a2adf1f81488829760fa2ba03/op-batcher/batcher/driver.go#L537) for failed batch transactions. This modification ensures that the data contained in failed transactions will be resubmitted automatically.
   1. Most failures will be detected during the [`EstimateGas`](https://github.com/ethereum-optimism/optimism/blob/8f516faf42da416c02355f9981add3137a3db190/op-service/txmgr/txmgr.go#L266) call. However, under certain race conditions, failures may occur after the transaction has been included in a block.

These modifications aim to enhance the security and efficiency of the batch submission and processing pipeline, allowing for more flexible and customizable conditions while maintaining the integrity of the derived state.

> ## Dynamic Updates

Reviewer: ```suggestion
## Supporting Dynamic Updates to Inbox Address
```

### SystemConfig

> The `SystemConfig` is the source of truth for the address of inbox. It stores information about the inbox address and passes the information to L2 as well. 

Reviewer: We should describe how `initialize` works for the address (whether the current behavior is the same or not?).

Author: Added a section for `initialize`: https://github.com/ethstorage/specs/pull/3/commits/ebddbec5ecb447c1311f0b5f4fbaa3d0a93bfb6a

#### setBatchInbox

A new function `setBatchInbox` is introduced to the `SystemConfig` contract, enabling dynamic updates to the inbox:

```solidity
/// @notice Updates the batch inbox address. Can only be called by the owner.
/// @param _batchInbox New batch inbox address.

> function setBatchInbox(address _batchInbox) external onlyOwner {

Reviewer: What happens if `_batchInbox` is set to `0`?

Author: The 0 address will be used as inbox, which is also EOA.

Reviewer: I believe if `address` is 0, then op-node will be Rollup config's inbox to derive.  Please double check.

Author: Currently `ProcessSystemConfigUpdateLogEvent` doesn't support dynamic update for inbox, it only supports dynamic update for batcher address. It will be modified to handle inbox change.

    _setBatchInbox(_batchInbox);
}

/// @notice Updates the batch inbox address.
/// @param _batchInbox New batch inbox address.
function _setBatchInbox(address _batchInbox) internal {
    Storage.setAddress(BATCH_INBOX_SLOT, _batchInbox);

    bytes memory data = abi.encode(_batchInbox);
    emit ConfigUpdate(VERSION, UpdateType.BATCH_INBOX, data);
}

```

#### UpdateType

A new enum value `BATCH_INBOX` is added to the `UpdateType` enumeration.

```solidity
enum UpdateType {
    BATCHER,
    GAS_CONFIG,
    GAS_LIMIT,
    UNSAFE_BLOCK_SIGNER,
    BATCH_INBOX
}
```

### How `op-node` knows the canonical batch inbox

> We define the canonical batch inbox at L2 block `n` as the batch inbox at the L1 origin of L2 block `n`.

Reviewer: Unclear variables: two `n` are defined twice.

Author: The two `n` actually refers to the same one.

Author: What about the updated [version](https://github.com/ethstorage/specs/pull/3/commits/ea9b6b0d061f1d1d79b7c747b981b8e6d7fa3178)?

Under normal conditions, `op-node` knows the canonical batch inbox through the derivation pipeline:

> 1. The `L1Traversal` componenet first identifies the L1 `SystemConfig` changes while traversing the L1 block, via [`UpdateSystemConfigWithL1Receipts`](https://github.com/ethereum-optimism/optimism/blob/71928829ca7ece48152159daa1d231eac2df03b3/op-node/rollup/derive/l1_traversal.go#L78). This includes batcher and inbox changes.

Reviewer: ```suggestion
1. The `L1Traversal` component first identifies the L1 `SystemConfig` changes while traversing the L1 block, via [`UpdateSystemConfigWithL1Receipts`](https://github.com/ethereum-optimism/optimism/blob/71928829ca7ece48152159daa1d231eac2df03b3/op-node/rollup/derive/l1_traversal.go#L78).
```

   1. The [`ProcessSystemConfigUpdateLogEvent`](https://github.com/ethereum-optimism/optimism/blob/71928829ca7ece48152159daa1d231eac2df03b3/op-node/rollup/derive/system_config.go#L59) function will be modified to parse the newly added inbox change.
2. The `L1Retrieval` component then fetches the canonical batch inbox from the `L1Traversal` componenet and [pass](https://github.com/ethereum-optimism/optimism/blob/71928829ca7ece48152159daa1d231eac2df03b3/op-node/rollup/derive/l1_retrieval.go#L57) it to the `DataSourceFactory` component, via `OpenData` function, similar to how [`SystemConfig.BatcherAddr`](https://github.com/ethereum-optimism/optimism/blob/71928829ca7ece48152159daa1d231eac2df03b3/op-service/eth/types.go#L382) is handled.

> 3. The `FetchingAttributesBuilder` component is updated to incorporate the canonical batch inbox into the `DepositTx`, via [`L1InfoDeposit`](https://github.com/ethereum-optimism/optimism/blob/71928829ca7ece48152159daa1d231eac2df03b3/op-node/rollup/derive/l1_block_info.go#L263). This modified `DepositTx` is subsequently used to calculate the `SystemConfig` during L2 chain reorganization.

Reviewer: ```suggestion
3. The `FetchingAttributesBuilder` component is updated to incorporate the canonical batch inbox into the `DepositTx`, via [`L1InfoDeposit`](https://github.com/ethereum-optimism/optimism/blob/71928829ca7ece48152159daa1d231eac2df03b3/op-node/rollup/derive/l1_block_info.go#L263). This modified `DepositTx` is subsequently used to obtain the `SystemConfig` during L2 chain reorganization.
```

> During L2 reorganization, `op-node` knows the canonical batch inbox using the `SystemConfig` parameter of [`ResettableStage.Reset(context.Context, eth.L1BlockRef, eth.SystemConfig)`](https://github.com/ethereum-optimism/optimism/blob/71928829ca7ece48152159daa1d231eac2df03b3/op-node/rollup/derive/pipeline.go#L38) function.

Reviewer: ```suggestion
During L2 reorganization, `op-node` knows the canonical batch inbox using the `SystemConfig` parameter of [`ResettableStage.Reset(context.Context, eth.L1BlockRef, eth.SystemConfig)`](https://github.com/ethereum-optimism/optimism/blob/71928829ca7ece48152159daa1d231eac2df03b3/op-node/rollup/derive/pipeline.go#L38) function, where `SystemConfig` is derived from the `DepositTx` of the corresponding L2 block.
```

3. The `FetchingAttributesBuilder` component is updated to incorporate the canonical batch inbox into the `DepositTx`, via [`L1InfoDeposit`](https://github.com/ethereum-optimism/optimism/blob/71928829ca7ece48152159daa1d231eac2df03b3/op-node/rollup/derive/l1_block_info.go#L263). This modified `DepositTx` is subsequently used to obtain the `SystemConfig` during L2 chain reorganization.

During L2 reorganization, `op-node` knows the canonical batch inbox using the `SystemConfig` parameter of [`ResettableStage.Reset(context.Context, eth.L1BlockRef, eth.SystemConfig)`](https://github.com/ethereum-optimism/optimism/blob/71928829ca7ece48152159daa1d231eac2df03b3/op-node/rollup/derive/pipeline.go#L38) function, where `SystemConfig` is derived from the `DepositTx` of the corresponding L2 block.

### How `op-batcher` knows canonical batch inbox

Immediately before submitting a new batch, `op-batcher` fetches the current inbox address from L1 and submits to that address. After the transaction is successfully included in L1 at block `N`, `op-batcher` verifies that the inbox address hasn't changed at block `N`. If the address has changed, it resubmits the batch to the new address.

## Upgrade

> Existing OP Stack instances need to upgrade the `SystemConfig` and set an inbox contract in order to use this feature.

Reviewer: I don't think this step is necessary to use the inbox contract: we can use the old `SystemConfig`.  This is only needed when performing `dynamic upgrade` (or migration).

Author: If the contract is not updated, there will be no `ConfigUpdate(VERSION, UpdateType.BATCH_INBOX, data)` event.

/// @param _batchInbox New batch inbox address.
function _setBatchInbox(address _batchInbox) internal {
    Storage.setAddress(BATCH_INBOX_SLOT, _batchInbox);

    bytes memory data = abi.encode(_batchInbox);
    emit ConfigUpdate(VERSION, UpdateType.BATCH_INBOX, data);
}

```

#### initialize

The `SystemConfig` now emits an event when the inbox is initialized, while retaining its existing support for inbox configuration during initialization.

```solidity
function initialize(
    address _owner,
    uint32 _basefeeScalar,
    uint32 _blobbasefeeScalar,
    bytes32 _batcherHash,
    uint64 _gasLimit,
    address _unsafeBlockSigner,
    ResourceMetering.ResourceConfig memory _config,
    address _batchInbox,
    SystemConfig.Addresses memory _addresses
)
    public
    initializer
{
    ...
    // Storage.setAddress(BATCH_INBOX_SLOT, _batchInbox);
    // initialize inbox by `_setBatchInbox` so that an event is emitted.
    _setBatchInbox(_batchInbox);
    ...
}
```

#### UpdateType

A new enum value `BATCH_INBOX` is added to the `UpdateType` enumeration.

```solidity
enum UpdateType {
    BATCHER,
    GAS_CONFIG,
    GAS_LIMIT,
    UNSAFE_BLOCK_SIGNER,
    BATCH_INBOX
}
```

> ### L1Block

Reviewer: I would suggest following the description of custom gas token:

```
L1Block
The L1Block contract is upgraded with a permissioned function setGasPayingToken that is used to set information about the gas paying token. The DEPOSITOR_ACCOUNT MUST be the only account that can call the setter function. This setter is differentiated from the setL1BlockValues functions because the gas paying token is not meant to be dynamic config whereas the arguments to setL1BlockValues are generally dynamic in nature.

Any L2 contract that wants to learn the address of the gas paying token can call the getter on the L1Block contract.
```

> The `L1Block` stores Layer 1 (L1) related information on Layer 2 (L2). It is extended to also store the inbox address.

Reviewer: ```suggestion
The `L1Block` stores Layer 1 (L1) related information on Layer 2 (L2). It is extended to store the dynamic inbox address.
```

3. The `FetchingAttributesBuilder` component is updated to incorporate the canonical batch inbox into the `DepositTx`, via [`L1InfoDeposit`](https://github.com/ethereum-optimism/optimism/blob/71928829ca7ece48152159daa1d231eac2df03b3/op-node/rollup/derive/l1_block_info.go#L263). This modified `DepositTx` is subsequently used to obtain the `SystemConfig` during L2 chain reorganization.

During L2 reorganization, `op-node` knows the canonical batch inbox using the `SystemConfig` parameter of [`ResettableStage.Reset(context.Context, eth.L1BlockRef, eth.SystemConfig)`](https://github.com/ethereum-optimism/optimism/blob/71928829ca7ece48152159daa1d231eac2df03b3/op-node/rollup/derive/pipeline.go#L38) function, where `SystemConfig` is derived from the `DepositTx` of the corresponding L2 block.

### How `op-batcher` knows canonical batch inbox

Immediately before submitting a new batch, `op-batcher` fetches the current inbox address from L1 and submits to that address. After the transaction is successfully included in L1 at block `N`, `op-batcher` verifies that the inbox address hasn't changed at block `N`. If the address has changed, it resubmits the batch to the new address.

> ## Upgrade

Reviewer: I would recommend to follow https://github.com/ethstorage/specs/blob/7248e23a6c93053a5cae5906c8521f30b55165c0/specs/experimental/custom-gas-token.md#upgrade for a better description of upgrade.

# Inbox Contract

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**

> - [Inbox Contract](#inbox-contract)

Reviewer: Following custom gas token, this line should be removed?

    }
    if err := binary.Write(w, binary.BigEndian, info.BlobBaseFeeScalar); err != nil {
      return nil, err
    }
    if err := binary.Write(w, binary.BigEndian, info.SequenceNumber); err != nil {
      return nil, err
    }
    if err := binary.Write(w, binary.BigEndian, info.Time); err != nil {
      return nil, err
    }
    if err := binary.Write(w, binary.BigEndian, info.Number); err != nil {
      return nil, err
    }
    if err := solabi.WriteUint256(w, info.BaseFee); err != nil {
      return nil, err
    }
    blobBasefee := info.BlobBaseFee
    if blobBasefee == nil {
      blobBasefee = big.NewInt(1) // set to 1, to match the min blob basefee as defined in EIP-4844
    }
    if err := solabi.WriteUint256(w, blobBasefee); err != nil {
      return nil, err
    }
    if err := solabi.WriteHash(w, info.BlockHash); err != nil {
      return nil, err
    }
    // ABI encoding will perform the left-padding with zeroes to 32 bytes, matching the "batcherHash" SystemConfig format and version 0 byte.
    if err := solabi.WriteAddress(w, info.BatcherAddr); err != nil {
      return nil, err
    }
    // This is where marshalBinaryInboxFork differs from marshalBinaryEcotone

>     if err := solabi.WriteHash(w, info.BatchInbox); err != nil {

Reviewer: Should here be `WriteAddress` (the same as batcherHash).

This would then cause the proxy implementation set transactions to set an incorrect
implementation address, resulting in a bricked contract. The calldata is not generated
dynamically to enable deterministic upgrade transactions across all networks.

The proxy upgrade transactions are from `address(0)` because the `Proxy` implementation
considers `address(0)` to be an admin. Going straight to the `Proxy` guarantees that
the upgrade will work because there is no guarantee that the `Proxy` is owned by the
`ProxyAdmin` and going through the `ProxyAdmin` would require stealing the identity
of its owner, which may be different on every chain. That would require adding L2
RPC access to the derivation pipeline and make the upgrade transactions non deterministic.

### L1Block Deployment

- `from`: `0x4210000000000000000000000000000000000002`
- `to`: `null`
- `mint`: `0`
- `value`: `0`
- `gasLimit`: TODO
- `data`: TODO
- `sourceHash`: TODO

### L1Block Proxy Update

- `from`: `0x0000000000000000000000000000000000000000`
- `to`: `0x4200000000000000000000000000000000000015`
- `mint`: `0`
- `value`: `0`
- `gasLimit`: `50,000`
- `data`: TODO
- `sourceHash`: TODO

### SystemConfig Upgrade

> Finally, upgrade the `SystemConfig` and set an inbox contract on L1.

Reviewer: ```suggestion
Finally, to dynamically change the inbox address, `SystemConfig` on L1 will be upgraded to accept a new inbox address.
```
